From d9f029a501d9a35c06ae3dc60177b89c19acf629 Mon Sep 17 00:00:00 2001
From: Priyajit Ghosh <prghosh@habana.ai>
Date: Wed, 23 Oct 2024 20:25:58 +0300
Subject: [PATCH] [FSW-9512] update current security fw version to psoc register

This patch adds support to write security fw version
into  psoc security block CURRENT_FW_VERSION before
going to ppboot.

Change-Id: I6de2c86369c69b2d287c12cc8b2136df1155088f
Signed-off-by: Priyajit Ghosh <prghosh@habana.ai>
---

diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
index f9b56e1..245687b 100644
--- a/arch/fs1/arc/dev_security.c
+++ b/arch/fs1/arc/dev_security.c
@@ -13,6 +13,7 @@
 #include <status.h>
 #include <dev_security.h>
 #include <security.h>
+#include <boot_desc.h>
 
 /*
  * This key hash is calculated on the Public Key Properties -
@@ -103,6 +104,22 @@
 	boot_rom_global.sec_data.sec_ver = fw_version_sec;
 }
 
+void security_report_fw_version(enum hl_fw_manifest_type manifest_idx)
+{
+	struct hl_fw_boot_ptrs *boot_ptr = NULL;
+	uint32_t boot_ptr_addr = 0;
+
+	boot_ptr_addr = boot_desc_get_boot_pointer(manifest_idx);
+
+	if (boot_ptr_addr == 0)
+		return;
+
+	boot_ptr = (struct hl_fw_boot_ptrs *)boot_ptr_addr;
+
+	regmap_writel(boot_ptr->sec_ver,
+		IO0_H0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_CURRENT_FW_VERSION));
+}
+
 void security_load_rom_key_hash()
 {
 	uint8_t key_idx, dword;
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index 0c4a18c..f0a2e3e 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -30,6 +30,7 @@
 void boot_rom_eom_overrides(void);
 void security_load_fw_version(void);
 void security_load_rom_key_hash(void);
+void security_report_fw_version(enum hl_fw_manifest_type manifest_idx);
 
 extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
 
diff --git a/common/boot_flow.c b/common/boot_flow.c
index 2f4fc52..d90b377 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -114,7 +114,7 @@
 	if (boot_desc_get_img_desc(man_idx, HL_FW_IMG_PPBOOT, &desc, 0) == BOOT_ROM_SUCCESS) {
 		/* No need to verify as image validation is done in previous boot */
 		if (desc.start_addr) {
-			/* TODO: Report security image fw version */
+			security_report_fw_version(man_idx); /*Reporting current SVN*/
 			ppboot = (void *)((uint32_t)boot_rom_get_arc_addr(desc.start_addr));
 			boot_rom_status_update(PROGRESS_PPBOOT_START);
 			boot_rom_putc('p');
@@ -158,7 +158,7 @@
 	/* TODO: TPM PCR extend */
 
 	if (desc.start_addr) {
-		/* TODO: Report security image fw version */
+		security_report_fw_version(man_idx); /*Reporting current SVN*/
 		ppboot = (void *)((uint32_t)boot_rom_get_arc_addr(desc.start_addr));
 		boot_rom_status_update(PROGRESS_PPBOOT_START);
 		boot_rom_putc('p');
